lib/supavisor/handlers/proxy/handler.ex

```diff
@@ -0,0 +1,135 @@
+defmodule Supavisor.Handlers.Proxy.Handler do
+  @moduledoc false
+
+  require Logger
+
+  @behaviour :ranch_protocol
+  @behaviour :gen_statem
+
+  alias Supavisor.{
+    Helpers,
+    Protocol.Server,
+    Handlers.Proxy.Db,
+    Handlers.Proxy.Client
+  }
+
+  alias Supavisor.HandlerHelpers, as: HH
+
+  @sock_closed [:tcp_closed, :ssl_closed]
+  @proto [:tcp, :ssl]
+
+  @impl true
+  def start_link(ref, transport, opts) do
+    pid = :proc_lib.spawn_link(__MODULE__, :init, [ref, transport, opts])
+    {:ok, pid}
+  end
+
+  @impl true
+  def callback_mode, do: [:handle_event_function]
+
+  @impl true
+  def init(_), do: :ignore
+
+  def init(ref, trans, opts) do
+    Process.flag(:trap_exit, true)
+    Helpers.set_max_heap_size(90)
+
+    {:ok, sock} = :ranch.handshake(ref)
+    :ok = trans.setopts(sock, active: true)
+    Logger.debug("ClientHandler is: #{inspect(self())}")
+
+    data = %{
```

Comment:
I still think that `defstruct` for state would be useful.

Reply:
100%